Fix top_file_merging_strategy warning if env_order is set

bdrung opened the pull request:
### What does this PR do?

According to the documentation, the `env_order` config option allows one to explicitly define the order in which top files are evaluated, when `top_file_merging_strategy` is set to `merge`, and no environment is specified for a highstate.

Despite honoring the environment order specified in `env_order` in `BaseHighState._get_envs()`, salt still prints a warning recommending setting env_order. Thus silence the warning when `env_order` is set.

In addition, remove duplicate code and simplify `_get_envs()` by using list comprehensions.

### What issues does this PR fix or reference?

This PR fixes #29104

### Tests written?

No

bdrung committed:
Remove duplicate client_envs variable definitions

self.client.envs() is called twice and assigned to client_envs and
client_env_list. Remove one of both variables since they are not
modified before usage.

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

bdrung committed:
Fix top_file_merging_strategy warning if env_order is set

According to the documentation, the env_order config option allows one
to explicitly define the order in which top files are evaluated, when
top_file_merging_strategy is set to merge, and no environment is
specified for a highstate.

Despite honoring the environment order specified in env_order in
BaseHighState._get_envs(), salt still prints a warning recommending
setting env_order. Thus silence the warning when env_order is set.

fixes #29104

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

terminalmage commented on salt/state.py:
```diff
@@ -2507,22 +2508,16 @@ def _get_envs(self):
                          if x not in envs])
         env_order = self.opts.get('env_order', [])
+        # Remove duplicates
+        env_order = list(collections.OrderedDict.fromkeys(env_order))
```
May I ask why we're using this approach? There are many less-complicated ways of removing duplicates. Not to mention the fact that this method seems to be incredibly inefficient:

```python
>>> import timeit
>>> prep = 'env_order = "[\'foo\', \'bar\', \'foo\', \'baz\']"; import collections'
>>> t1 = timeit.Timer('list(set(env_order))', prep)
>>> t2 = timeit.Timer('list(collections.OrderedDict.fromkeys(env_order))', prep)
>>> t1.repeat()
[1.3065440654754639, 1.2755541801452637, 1.2373528480529785]
>>> t2.repeat()
[22.5503408908844, 25.115081071853638, 24.109618186950684]
```

bdrung replied:
list(set(env_order)) does not preserve the order. The used version is the shortest one for removing duplicates with preserving the order.

terminalmage replied:
OK, but you can still iterate over a list or use a list comprehension and not incur all that additional overhead.

cachedout replied:
@bdrung Did you get a chance to follow-up here?

bdrung replied:
There are multiple variants to remove duplicates:
```python
>>> prep = 'env_order = "[\'foo\', \'bar\', \'foo\', \'baz\']"; import collections'
>>> import timeit
>>> t1=timeit.Timer('sorted(set(env_order), key=lambda x: env_order.index(x))', prep)
>>> t1.repeat()
[4.158413887023926, 3.958709955215454, 3.9590039253234863]
>>> t1 = timeit.Timer('list(set(env_order))', prep)
>>> t1.repeat()
[0.9367499351501465, 0.903162956237793, 0.9018638134002686]
>>> t2 = timeit.Timer('list(collections.OrderedDict.fromkeys(env_order))', prep)
>>> t2.repeat()
[16.622393131256104, 16.300525903701782, 16.367045164108276]
>>> t3=timeit.Timer('sorted(set(env_order), key=lambda x: env_order.index(x))', prep)
>>> t3.repeat()
[4.10307502746582, 4.041788101196289, 4.156888961791992]
>>> t4=timeit.Timer('[v for (i, v) in enumerate(env_order) if v not in env_order[0:i]]', prep)
>>> t4.repeat()
[2.9662280082702637, 2.8464999198913574, 2.7960379123687744]
>>> t5=timeit.Timer('dummy = set(); [(v, dummy.add(v))[0] for v in env_order if v not in dummy]', prep)
>>> t5.repeat()
[2.429576873779297, 2.391568899154663, 2.5268590450286865]
```
Which of the order preserving variants do you prefer?

terminalmage replied:
I think that a variant on the last method which uses a set to store which are members and which are not, would be best:

```python
>>> import timeit
>>> prep = 'env_order = "[\'foo\', \'bar\', \'foo\', \'baz\']"'
>>> timeit.Timer('dummy = set(); [(v, dummy.add(v))[0] for v in env_order if v not in dummy]', prep).repeat()
[3.2361321449279785, 3.179112195968628, 3.1725778579711914]
>>> timeit.Timer('members = set(); env_order = [x for x in env_order if not (x in members or members.add(x))]', prep).repeat()
[1.925464153289795, 1.9767980575561523, 1.8665900230407715]
```

This is superior to the ``t5`` example, I think, because it avoids both the instantiation of a tuple, and the reference to the index of the tuple. It also avoids using ``set.add`` for each iteration.

bdrung replied:
Okay. I updated the PR to use the last proposed variant.

bdrung committed:
Simplify _get_envs() by using list comprehensions

final_list creates an ordered intersection of env_order and client_envs
without duplicates. Instead of creating a intersection set, remove the
duplicates in env_order and generate the ordered intersection with a
simple list comprehension.

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

cachedout commented:
@terminalmage Can you please re-review?

terminalmage commented on salt/state.py:
```diff
@@ -2591,7 +2586,7 @@ def get_tops(self):
                     log.debug('No contents loaded for env: {0}'.format(saltenv))
 
-            if found > 1 and merging_strategy == 'merge':
+            if found > 1 and merging_strategy == 'merge' and not self.opts.get('env_order', None):
```
The ``.get()`` is probably unnecessary, since the opts will always contain an ``env_order`` key, and the default value is an empty list. But this is a minor nitpick and it shouldn't prevent this from being merged.
